broadinstitute/pearl
# [JN-539] Form JSON editor validation

This detects a bunch of different cases:

Invalid JSON:
<img width="759" alt="Screenshot 2023-08-30 at 2 18 08 PM" src="https://github.com/broadinstitute/juniper/assets/7257391/e1993939-4467-4ef1-9955-81c5078389f1">

Invalid question template references:
<img width="767" alt="Screenshot 2023-08-30 at 2 17 58 PM" src="https://github.com/broadinstitute/juniper/assets/7257391/7eabacbe-32a2-445e-8814-51c4b499d68a">

Elements with missing types:
<img width="762" alt="Screenshot 2023-08-30 at 2 18 31 PM" src="https://github.com/broadinstitute/juniper/assets/7257391/4cc5e328-e659-4f74-b60b-98be83f5e1d0">

Elements with missing names:
<img width="761" alt="Screenshot 2023-08-30 at 2 18 46 PM" src="https://github.com/broadinstitute/juniper/assets/7257391/dbcc5d9b-458d-43cb-83f8-4e75584bd157">

Elements with duplicate names:
<img width="761" alt="Screenshot 2023-08-30 at 2 19 17 PM" src="https://github.com/broadinstitute/juniper/assets/7257391/fd48ccef-5818-40e1-bc92-9dbe32054d6d">

Forms without pages:
<img width="766" alt="Screenshot 2023-08-30 at 2 21 18 PM" src="https://github.com/broadinstitute/juniper/assets/7257391/9f1d59c7-054d-46aa-82fd-3fbf8832b65e">

Pages without elements:
<img width="761" alt="Screenshot 2023-08-30 at 2 21 29 PM" src="https://github.com/broadinstitute/juniper/assets/7257391/b5512714-dad3-48ca-a639-5ae5c30b99ec">

Panels without elements:
<img width="765" alt="Screenshot 2023-08-30 at 2 21 37 PM" src="https://github.com/broadinstitute/juniper/assets/7257391/103cec52-36b1-4041-bcd3-8b1bf37658c8">


TO TEST:

* Try out all of the above test cases
* Make sure that you can correct the errors and everything still saves correctly
* Make sure that the Designer and Preview tabs are disabled while there are errors, and that the Save button is disabled
* Make sure that you don't encounter any of these errors while playing around in the Designer tab (there are enough guardrails in place, so you shouldn't be able to trigger any of these errors)
